src/librustc_ast_lowering/expr.rs

```diff
@@ -175,37 +207,14 @@ impl<'hir> LoweringContext<'_, 'hir> {
                     maybe_expr,
                 )
             }
-            ExprKind::Paren(ref ex) => {
-                let mut ex = self.lower_expr_mut(ex);
-                // Include parens in span, but only if it is a super-span.
-                if e.span.contains(ex.span) {
-                    ex.span = e.span;
-                }
-                // Merge attributes into the inner expression.
-                let mut attrs = e.attrs.clone();
-                attrs.extend::<Vec<_>>(ex.attrs.into());
-                ex.attrs = attrs;
-                return ex;
-            }
-
             ExprKind::Yield(ref opt_expr) => self.lower_expr_yield(e.span, opt_expr.as_deref()),
-
             ExprKind::Err => hir::ExprKind::Err,
-
-            // Desugar `ExprForLoop`
-            // from: `[opt_ident]: for <pat> in <head> <body>`
-            ExprKind::ForLoop(ref pat, ref head, ref body, opt_label) => {
-                return self.lower_expr_for(e, pat, head, body, opt_label);
-            }
             ExprKind::Try(ref sub_expr) => self.lower_expr_try(e.span, sub_expr),
-            ExprKind::Mac(_) => panic!("Shouldn't exist here"),
-        };
 
-        hir::Expr {
-            hir_id: self.lower_node_id(e.id),
-            kind,
-            span: e.span,
-            attrs: e.attrs.iter().map(|a| self.lower_attr(a)).collect::<Vec<_>>().into(),
+            ExprKind::ForLoop(..) | ExprKind::Paren(_) => {
+                span_bug!(e.span, "Should be handled by `lower_expr`")
+            }
```

Comment:
This seems arbitrary; how was this chosen? (Also leave a comment about it in the code.)

Reply:
@oli-obk may be able to answer this better. My guess is that this was done to heuristically reduce number of calls to `ensure_sufficient_stack`. Though at at least unrolling of the `ExprKind::Paren` structure is still recursive so not sure if it makes the sense to extract it anymore.